src/atlas/interpolation/method/binning/Binning.cc

```diff
@@ -0,0 +1,186 @@
+/*
+ * (C) Crown Copyright 2024 Met Office
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+
+#include "atlas/functionspace/NodeColumns.h"
+#include "atlas/functionspace/StructuredColumns.h"
+#include "atlas/grid.h"
+#include "atlas/interpolation/Interpolation.h"
+#include "atlas/interpolation/method/binning/Binning.h"
+#include "atlas/interpolation/method/MethodFactory.h"
+#include "atlas/runtime/Trace.h"
+
+#include "eckit/config/LocalConfiguration.h"
+#include "eckit/linalg/SparseMatrix.h"
+#include "eckit/linalg/Triplet.h"
+#include "eckit/mpi/Comm.h"
+
+
+namespace atlas {
+namespace interpolation {
+namespace method {
+
+
+namespace {
+
+MethodBuilder<Binning> __builder("binning");
+
+}
+
+
+Binning::Binning(const Config& config) : Method(config) {
+  const auto* ptr_config = dynamic_cast<const eckit::LocalConfiguration*>(&config);
+  interpAncillaryScheme_ = ptr_config->getSubConfiguration("ancillary_scheme");
+  halo_exchange_ = ptr_config->getBool("halo_exchange", true);
+  adjoint_ = ptr_config->getBool("adjoint", false);
+}
+
+
+void Binning::do_setup(const Grid& source,
+                       const Grid& target,
+                       const Cache&) {
+  ATLAS_NOTIMPLEMENTED;
+}
+
+
+void Binning::do_setup(const FunctionSpace& source,
+                       const FunctionSpace& target) {
+  ATLAS_TRACE("atlas::interpolation::method::Binning::do_setup()");
+
+  using Index = eckit::linalg::Index;
+  using Triplet = eckit::linalg::Triplet;
+  using SMatrix = eckit::linalg::SparseMatrix;
+
+  source_ = source;
+  target_ = target;
+
+  if (target_.size() == 0) {
+    return;
+  }
+
+  // enabling or disabling the halo exchange
+  this->allow_halo_exchange_ = halo_exchange_;
+  // enabling or disabling the adjoint operation
+  this->adjoint_ = adjoint_;
```

Comment:
This sets `adjoint_` to `adjoint_` (no-op). Probably not intentional?

Reply:
I implemented the adjoint operation as _optional_ -- _i.e._ if the operation is needed by the client, the client can set the flag to `true` and the setup of the adjoint operation will be carried out.
Do you see potential risks in this approach? ...
<br>

Reply:
Thanks I understand now. The problem is that the base class's `bool Method::adjoint_` which you were trying to set is private, and instead you were setting an overriding `bool Binning::adjoint_`.  I now made `bool Method::adjoint_`  protected (not private), and removed the overriding values in [b09c2a5](https://github.com/ecmwf/atlas/pull/191/commits/b09c2a5971765e4b8c2a1947dc4ab7245c87c3fc) so that the intended values are now modified.
I wonder why this bug was  failure did not occur before. Possibly some test was missing?
Also, @odlomax , I noticed a similar `bool SphericalVector::adjoint_` value in "SphericalVector" which may or may not be intending to overwrite the former private `Method::adjoint_`. Could you please verify that all is as expected in this regard?

Reply:
@wdeconinck,
Yep, that why I did it. It'll be safe to change it to the base class version once it's accessible.

Reply:
@odlomax --> [d31023e](https://github.com/ecmwf/atlas/pull/191/commits/d31023ed05151b46713f38cb7915237fb0ef8372)